mlir/lib/Target/Cpp/TranslateToCpp.cpp

```diff
@@ -1570,6 +1570,10 @@ LogicalResult CppEmitter::emitType(Location loc, Type type) {
   }
   if (auto iType = dyn_cast<IndexType>(type))
     return (os << "size_t"), success();
+  if (auto sType = dyn_cast<emitc::SizeTType>(type))
+    return (os << "size_t"), success();
+  if (auto sType = dyn_cast<emitc::SignedSizeTType>(type))
+    return (os << "ssize_t"), success();
```

Comment:
So far we were trying to follow the convention documented in https://github.com/llvm/llvm-project/blob/59c6bd156cc8b42758ce90909615748e21c6eee2/mlir/docs/Dialects/emitc.md, which at some point states `Else the generated code is compatible with C99.` As far as I know, this doesn't hold if `ssize_t` is emitted as this is included in the POSIX `<sys/types.h>` header but isn't strict C99. So the minimum is to update the docs accordingly. However, this introduces another include as discussed on [Discord](https://discord.com/channels/636084430946959380/1209151815007080468/1242366862549848086) (we should really try to move longer discussions to discourse though :smile:).

Furthermore, how do we handle if the emitted code should be C++?

Reply:
Great point. I keep worrying about the fact that `ssize_t` can only represent -1 where we'll need another type for `index`-type values below -1. How about using [`ptrdiff_t`](https://en.cppreference.com/w/c/types/ptrdiff_t)?

Reply:
Hm, this is my fault, I though `ssize_t` was standardized.  How important is that to you?

We use `ssize_t` in lowerings that need to interpret `index` as `signed` (like `index.divs`). 

Just brainstorming, we could emit this as `intmax_t`, `ptrdiff_t` (C) or `std::make_signed<std::size_t>` (C++11). 
([cppreference says](https://en.cppreference.com/w/cpp/types/ptrdiff_t): `For char arrays shorter than PTRDIFF_MAX, std::ptrdiff_t acts as the signed counterpart of std::size_t`)

Reply:
Me too, don't worry. I thought those two types came together, indeed they don't. 

`intmax_t` will be able to represent any value `index` can, `ptrdiff_t` seems more geared towards pointer arithmetics. On the other hand, if we use `size_t` for the unsigned type, casts from `intmax_t` to `size_t` aren't guaranteed to preserve the value (`intmax_t` may be much larger than `size_t`), whereas

> For char arrays shorter than [PTRDIFF_MAX](https://en.cppreference.com/w/c/types/limits), ptrdiff_t acts as the signed counterpart of [size_t](https://en.cppreference.com/w/c/types/size_t)

also per the MLIR doc:

> The index type models target-specific values of pointer width, like `intptr_t`.

so we don't need anything wider than `ptrdiff_t`. Should we use `intmax_t` for the signed type, we'd need `uintmax_t` for the unsigned type.

Thoughts?

Reply:
No worries :) We don't need to force ourselves to emit strict C99, but we need to at least document what IR will result in what kind of code.

Since `intmax_t` and `uintmax_t` are available since C99 and C++11, and since fixed width integer types are emitted anyway (thus we don't enforce to include yet another header) this is probably what I would prefer if that works for your lowering. Adding those two as explicit types to EmitC seems reasonable as well as `ptrdiff_t`.

Reply:
The only requirement is that values are preserved when going back and forth between the signed and the unsigned representative for the index type (these casts are auto-inserted depending on which signedness is needed by the op). This looks possible for `intmax_t` and `uintmax_t`.

Reply:
So here's what I did in the end: we'll keep `size_t`, `ssize_t` in EmitC with this note that `ssize_t` goes out of C99. We also add `ptrdiff_t` as a data type, and conversion functions to go back and forth between types with different signedness and same bitwidth.

This enables passes to make abstraction of which data type is actually used (and just have a data type that accurately represents `index`), while also providing them with explicit C-world data types they can use (e.g. a pass will want to explicitly use `size_t` if they're calling `sizeof`).